### PR TITLE
[luci-interpreter] Rework observing functionality support

### PR DESCRIPTION
**Pull request by s-barannikov**
Draft PR #1980.

Add a layer of abstraction to avoid dependency of runtime entities on IR entities.

ONE-DCO-1.0-Signed-off-by: Sergei Barannikov <s.barannikov@samsung.com>

### CODE REVIEW

File: compiler/luci-interpreter/src/core/Hook.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+#ifndef LUCI_INTERPRETER_CORE_HOOK_H
+#define LUCI_INTERPRETER_CORE_HOOK_H
+
+namespace luci_interpreter
+{
+
+// Used at execution stage to tell the interpreter that the runtime state has changed in some way.
+class Hook
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/core/Hook.h:24]: Please suggest a better name.

**Comment by s-barannikov** [comment]
@jinevening 
Could you check that quantizator is still working? I can't run the tests because I have newer version of Ubuntu, which lacks outdated TensorFlow python packages.

**Comment by jinevening** [comment]
> Could you check that quantizator is still working? I can't run the tests because I have newer version of Ubuntu, which lacks outdated TensorFlow python packages.

`record-minmax-conversion-test` runs without error.

BTW, the test runs python on `venv`, so it should be able to run on higher versions of Ubuntu (does it depend on Ubuntu version?).

**Comment by s-barannikov** [comment]
@jinevening 
Strictly speaking, it depends on Python version.
Ubuntu 20.04 comes with Python 3.8.2 onboard, which has  2.2.0rc1, 2.2.0rc2, 2.2.0rc3, 2.2.0rc4, 2.2.0 TensorFlow package versions.
Downgrading python is not a solution for me.

File: compiler/luci-interpreter/src/core/Hook.h (hunk 1 above)
Comment (jinevening) [reply compiler/luci-interpreter/src/core/Hook.h:24]: How about Notifier?

This class notifies an event (change of runtime state according to your comment) to the interpreter.

Comment (s-barannikov) [reply compiler/luci-interpreter/src/core/Hook.h:24]: > How about Notifier?

I also considered this option. `Notifier` sounds too ambigous. I can't say that `Hook` is more specific though :)

**Comment by jinevening** [comment]
I'm curious if we can expose `_node_to_tensor` (using a getter), so that an observer can use it.

E.g., If we allow an observer to access `_node_to_tensor`, the observer will be able to do more things (e.g., getting tensor data of the node's inputs).

File: compiler/luci-interpreter/src/core/Hook.h (hunk 1 above)
Comment (jinevening) [reply compiler/luci-interpreter/src/core/Hook.h:24]: `EventNotifier` sounds good to me :)

**Comment by seanshpark** [comment]
> Ubuntu 20.04 comes with Python 3.8.2 onboard, which has 2.2.0rc1, 2.2.0rc2, 2.2.0rc3, 2.2.0rc4, 2.2.0 TensorFlow package versions.

Out default TF is 1.13.1

**Comment by s-barannikov** [comment]
@seanshpark AFAIK TensorFlow versions are backward compatible. Our are there any issues?

**Comment by seanshpark** [comment]
>TensorFlow versions are backward compatible. Our are there any issues?

I don't know.  You wrote

> I can't run the tests because I have newer version of Ubuntu, which lacks outdated TensorFlow python packages.

What is your problem?

**Comment by s-barannikov** [comment]
@seanshpark I can only install 2.2.0 version and its release candidates. 1.13 is unavailable in the repo.
I suppose 2.2.0 would also work well for the purpose of testing, because versions shold be backward compatible, but I did not try.
We could at least make the version configurable via some cmake option. Right now it is hardcoded in requirements.txt.

File: compiler/luci-value-test/requirements.txt [hunk 2]
```diff
@@ -1,2 +1,2 @@
-tensorflow==1.13.1
+tensorflow==2.2.0
```
Comment (jinevening) [inline compiler/luci-value-test/requirements.txt:1]: This change seems not relevant to this PR.

Comment (s-barannikov) [reply compiler/luci-value-test/requirements.txt:1]: Oh, sorry, I committed this accidently. I will revert it

**Comment by seanshpark** [comment]
> I can only install 2.2.0 version and its release candidates. 

So can you change your platform that has 1.13.1 ?

File: compiler/luci-value-test/requirements.txt (hunk 2 above)
Comment (jinevening) [reply compiler/luci-value-test/requirements.txt:1]: Now I saw the above comment. If tf version is backward compatible, it's ok to change the version.

**Comment by s-barannikov** [comment]
@seanshpark Yes, I can. But we can't force external contributors to do the same.

**Comment by seanshpark** [comment]
> Yes, I can. But we can't force external contributors to do the same.

My point is our target TF is 1.13.1 and we should work on a platform that supports this product.
Do you mean that we should support TF 2.2 for Ubuntu 20.04 users?

File: compiler/luci-value-test/requirements.txt (hunk 2 above)
Comment (seanshpark) [reply compiler/luci-value-test/requirements.txt:1]: I don't think this is about backward compatibility. Please remove this change.
And this file is different from other files.

**Comment by s-barannikov** [comment]
@seanshpark I don't know if it is 100% working solution. It shold be, but who knows.
It seems that it would be better to install the required version (1.13), but I don't know how to do that without downgrading OS or Python.

**Comment by seanshpark** [comment]
>  downgrading OS or Python.

Have you tried docker with Ubuntu 18.04 ?

**Comment by s-barannikov** [comment]
> > downgrading OS or Python.
> 
> Have you tried docker with Ubuntu 18.04 ?

No. I've never tried building the project in docker. I will try, thank you for suggestion.

File: compiler/luci-interpreter/include/luci_interpreter/Interpreter.h [hunk 3]
```diff
@@ -59,8 +59,14 @@ class Interpreter
...
-  std::unique_ptr<class TensorMap> _tensor_map;
-  std::unique_ptr<class KernelMap> _kernel_map;
+  std::unique_ptr<class TensorMap> _node_to_tensor;
+
+  // Kernels, in execution order.
+  std::vector<std::unique_ptr<class Kernel>> _kernels;
+
+  // Observer functionality support.
+  std::unique_ptr<struct RuntimeToIR> _runtime_to_ir;
+  std::unique_ptr<class EventNotifier> _event_notifier;
```
Comment (binarman) [inline compiler/luci-interpreter/include/luci_interpreter/Interpreter.h:69]: I didn't get why you need `RuntimeToIR` structure and `EventNotifier`, could you clarify?

Comment (s-barannikov) [reply compiler/luci-interpreter/include/luci_interpreter/Interpreter.h:69]: It is in the description. To make runtime entities independent of IR.
Please see the draft for details.

Comment (binarman) [reply compiler/luci-interpreter/include/luci_interpreter/Interpreter.h:69]: My question was not about purpose of these members, but about purpose of separate structures.
From my point EventNotifier can be inlined, This applies to RuntimeToIR too.

Well, Let's return to this discussion, when everything is working =)

**Pull request merged**